## RepeatDefeaters - Utilities for unclassified consensus sequences

## Table of Contents
-   **Introduction**
    -   [Motivation](#motivation)
    -   [Key features](#key-features)

-   **Dependencies**
    -   [Databases](#DB)
    -   [Blast](#blast)
    -   [HMMER3](#hmmer3)
    -   [PfamScan](#pfamscan)

-   **Mechanism**
    -   [TE Accessions](#teaccs)
    -   [TE HMMs](#TEHMMs)

## Introduction

### Motivation

For recently sequenced non-model organisms, repeat discovery tools
often fail to classify a large portion of their repeats. These
unclassified repeats can be host genes that were duplicated, or TEs
that are just not present in the databases. Analyses of Transposable
Elements (TE) can be misleading if host genes are present.
RepeatDefeaters is a tool to further classify which repeats should
be considered host genes or TEs.

A critical task during repeat discovery is to accurately annotate
TEs. This can be challenging for new organisms that have few
references to compare with. RepeatDefeaters aims to provide an
easy-to-follow guideline on how to tackle these consensus sequences
that are difficult to classify automatically.

### Key features

RepeatDefeaters provides:
1. Utilities that work together to determine if your consensus
sequence of interest is related to TE activity.

## Usage

Usage:
```bash
nextflow run -c <parameter.config> [-profile <executor profile>] GroundB/RepeatDefeaters
```

where:
- `<parameter.config>` is a nextflow configuration file which provides
    paths to the input data (see [params.config template](params.config.TEMPLATE)).
    Alternatively parameters can be provided on the
    command-line using the `--parameter` notation (e.g., `--species_short_name <str>` ).
- `<executor profile>` is one of the preconfigured execution profiles
    (`uppmax`, `singularity_local`, `docker_local`). Alternatively,
    you can provide a custom configuration to configure this workflow
    to your execution environment. See [Nextflow Configuration](https://www.nextflow.io/docs/latest/config.html#scope-executor)
    for more details.

### Dependancies

- [Nextflow](Nextflow.io/): A workflow manager. It can be installed
    into a custom conda environment (recommended), or directly
    into your bin. A conda environment file (nextflow_conda-env.yml) is
    provided to create the environment.
- Package Manager: One of the following.
    - [Docker](https://www.docker.com/): A container platform
    - [Singularity](https://sylabs.io/singularity/):
    - [Conda](https://docs.conda.io/en/latest/miniconda.html):

##  Mechanism
In addition to stringent nucleotide searches against the TREP database, RepeatDefeaters finds TE proteins in provided consensus sequences, giving clues about the consensus sequence. After getting a hint, the consensus sequence can be compared with the literature and subsequently curated.

### TE Accessions
A list of Pfam accessions that might relate to TE activities was generated by running the `PfamTEs.sh` script under `/Scripts`. Results were put into `/Resources`. `PfamTEs.sh` scans transposon keywords in Pfam and extracts accession names accordingly. The output list is prefixed with Pfam.TE.accessions, followed by the Pfam release date (e.g., Pfam.TE.accessions_Release32). You can directly access the file under `/Resources` or run PfamTEs.sh yourself. Changes should and will be made on the automatically generated list since transposon keywords don't equal TE activities. A good example would be the transposon silencing proteins of the host.
Current deletions of entries:
```bash
grep -viw BTB Pfam.TE.accessions_Release32|grep -vi piwi > Pfam.TE.accessions_Release32.amended
```
### TE HMMs
TE protein alignments were made and subsequently built into HMMs in `Libraries`. TE protein sequences were sourced from TREP, Repbase, and REXDB. Due to copyright issues, these alignments are not put in this repository now. If you are interested in getting the alignment file, please get in touch.

### Contributing to RepeatDefeaters
I genuinely believe that a community-maintained database will serve the transposon research the best in the long-term. Active participation in annotating unclassified consensus sequences, establishing and optimizing profile-HMMs of TE proteins are very much needed. "RepeatDefeaters" should be a resource center that is reviewed and contributed by many, therefore the plural name.
For anyone wishing to help, join our [RepeatDefeaters Discord](https://discord.gg/SKatqNn).
