## RepeatDefeaters - Utilities for unclassified consensus sequences
### Table of Contents
-   **Introduction**
    -   [Motivation](#motivation)
    -   [Key features](#key-features)

-   **Dependencies**
    -   [Databases](#DB)
    -   [Blast](#blast)
    -   [HMMER3](#hmmer3)
    -   [PfamScan](#pfamscan)

-   **Mechanism**
    -   [TE Accessions](#teaccs)
    -   [TE HMMs](#TEHMMs)

### Introduction
#### Motivation
For non-model organisms that got sequenced recently, repeat discovery tools often fail to classify a large portion of their repeats. These unclassified repeats can be host genes that were duplicated or TEs that are just not present in the databases. It is misleading if host genes are included in analyses designed for TEs. But some studies had done that. That's why RepeatDefeaters provides a more transparent set of rules on what to be considered to be host or TEs.

One of the other critical tasks during repeat discovery is to accurately annotate transposable elements (TEs), which can be challenging for new organisms that have few references to compare with. RepeatDefeaters aims to provide an easy-to-follow guideline on how to tackle these consensus sequences that are difficult to be classified automatically.

Repeat annotations shall not stay as black boxes any longer.
#### Key features
RepeatDefeaters provides:
1. utilities that work together to determine If your consensus sequence of interest is related to TE activities
2. a more comprehensive collection of Hidden Markov Models (HMMs) for TE proteins
### Dependencies
#### Databases
Please download files from their FTP site.
```bash
#For Pfam release 32.0
wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam32.0/Pfam-A.hmm.gz
gunzip Pfam-A.hmm.gz
#You also need the Pfam-A.full.uniprot.gz for searching transposon related synonyms
wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam32.0/Pfam-A.full.uniprot.gz
gunzip Pfam-A.full.uniprot.gz
#Non-redundant TREP nucleotide sequences are required
wget http://botserv2.uzh.ch/kelldata/trep-db/downloads/trep-db_nr_Rel-19.fasta.gz
gunzip trep-db_nr_Rel-19.fasta.gz
```
#### Blast
blastx is used to predict amino acid sequences from consensus sequences of repeats.

blastn is used to search consensus sequences against the TREP nt database.

Version 2.6.0+

Link to the software: https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.6.0/
```bash
sudo apt-get update
sudo apt install ncbi-blast+
```

#### HMMER3
HMMER3 was used to build HMMs from TE protein alignments.

HMMER3 is also needed to search amino acid predictions against HMMs of TE proteins.

```bash
sudo apt install build-essential
wget http://eddylab.org/software/hmmer/hmmer-3.3.tar.gz
tar -xvf hmmer-3.3.tar.gz
cd /hmmer-3.3
./configure
make install
```

#### PfamScan
The perl script pfam_scan.pl from PfamScan1.5 was used to search predicted amino acid sequences to the Pfam database.
```bash
#Download PfamScan1.5
conda install -c bioconda pfam_scan
```
##  Mechanism
In addition to stringent nucleotide searches against the TREP database, RepeatDefeaters finds TE proteins in provided consensus sequences, giving clues about the consensus sequence. After getting a hint, the consensus sequence can be compared with the literature and subsequently curated.

### TE Accessions
A list of Pfam accessions that might relate to TE activities was generated by running the `PfamTEs.sh` script under `/Scripts`. Results were put into `/Resources`. `PfamTEs.sh` scans transposon keywords in Pfam and extracts accession names accordingly. The output list is prefixed with Pfam.TE.accessions, followed by the Pfam release date (e.g., Pfam.TE.accessions_Release32). You can directly access the file under `/Resources` or run PfamTEs.sh yourself. Changes should and will be made on the automatically generated list since transposon keywords don't equal TE activities. A good example would be the transposon silencing proteins of the host.
Current deletions of entries:
```bash
grep -viw BTB Pfam.TE.accessions_Release32|grep -vi piwi > Pfam.TE.accessions_Release32.amended
```
### TE HMMs
TE protein alignments were made and subsequently built into HMMs in `Libraries`. TE protein sequences were sourced from TREP, Repbase, and REXDB. Due to copyright issues, these alignments are not put in this repository now. If you are interested in getting the alignment file, please get in touch.

### Contributing to RepeatDefeaters
I genuinely believe that a community-maintained database will serve the transposon research the best in the long-term. Active participation in annotating unclassified consensus sequences, establishing and optimizing profile-HMMs of TE proteins are very much needed. "RepeatDefeaters" should be a resource center that is reviewed and contributed by many, therefore the plural name.
For anyone wishing to help, join our [RepeatDefeaters Discord](https://discord.gg/SKatqNn).
